### DISCUSSION TITLE
Brainstorming: board peripherals/sensors

### DISCUSSION THREAD

**Discussion by jessebraham** [esp-rs/esp-rust-board#2] [General]
During our brainstorming meeting today we determined that, at the very least, this board should include the following peripherals:

- Discrete LED (GPIO)
- User-defined button (GPIO)
- Accelerometer/gyroscope (I2C)
- Temperature/humidity sensor (I2C)
- Micro-SD card (SPI)

Please note that in its current state the README is not necessarily indicative of our plans. This was an initial draft and we have since had further internal discussions.

**Comment by uLipe** [comment]
It would be good to have an actuator output, maybe an open-drain mosfet with its gate attached to one of LEDC capable outputs.

**Reply by pedrominatel** [reply]
We can create a shield for motor control. What do you think?

**Reply by uLipe** [reply]
For motor control we could do something simple for Brushed-DC motors, but having the current sensor to provide the torque loop plus voltage sense to estimate the shaft velocity without requiring an encoder.

On BLDC side an L6234 plus an encoder / hall input would be great to perform voltage mode FoC control or something simpler like a 6-step commutation.

**Comment by suda-morris** [comment]
LCD which shares the same SPI bus with Micro-SD card (SPI), by different CS pin.

**Reply by jessebraham** [reply]
Due to space limitations I'm not sure we'll be able to fit an LCD on the board. We have not yet decided on the final form factor to my knowledge, but based on our previous internal discussions there likely won't be space. It is definitely a possibility though.

We were discussing maybe supporting expansion boards, or "shields", so if we do follow through with this idea it'd be easy enough to design an expansion board with an LCD.

**Comment by 0xjakob** [comment]
I personally found a Li-Po battery connector together with a charger quite handy with ESP-boards. Would make the power section slightly more complicated, though.

**Reply by jessebraham** [reply]
We discussed adding this in our internal meeting, however unfortunately with the current chip shortage this would likely add a non-trivial cost to the board for a feature that may not be used by the majority of developers. Nothing is set in stone yet though, so this may still be included.

**Comment by 0Grit** [comment]
I've designed a board like this in the past. Feel free to use it for inspiration. Slightly smaller than a business card.
 Obviously replace the application controller with a C3.

![image](https://user-images.githubusercontent.com/24318071/142559407-eb18504f-ae4e-4c26-9118-7fad5da31b57.png)

![image](https://user-images.githubusercontent.com/24318071/142559451-a9416a45-7611-4084-8176-62facefa2eef.png)

**Comment by 0Grit** [comment]
Definitely USB-C and please make use of USB-C's auxiliary mode which enables you to pass 14 signals and JTAG through the connector.

**Reply by jessebraham** [reply]
We definitely want USB-C on here as well!

I'm personally unfamiliar with the auxiliary mode, but thanks for the suggestion. @pedrominatel may know more about this and might be able to give you a more definitive answer.

**Reply by 0Grit** [reply]
https://en.wikipedia.org/wiki/USB-C#Debug_Accessory_Mode

It's not well known it seems.

**Reply by 0Grit** [reply]
Might be nice to follow a similar pinout as this board. It is the only other board I've seen taking advantage of Debug Accessory Mode.
https://media.digikey.com/pdf/Data%20Sheets/Maxim%20PDFs/MAXREFDES100HDK_Web.pdf
https://www.maximintegrated.com/en/design/reference-design-center/system-board/6312.html

**Reply by pedrominatel** [reply]
Since we have embedded JTAG on C3, I'm not sure if this mode will help us. I'm not aware of this mode, but we can check with our hardware team to see if we can use it on C3.

Thanks for this suggestion!

**Reply by 0Grit** [reply]
Even if you don't use it for JTAG you can use it for other signals like i2c, spi, etc.

**Comment by pedrominatel** [comment]
Thanks for all suggestions!

Last week we had a discussion about the development board and here is the summary:

**General**

- Hardware design tool: KiCad 5 (we can convert to KiCad 6 later)
- SoC: Bare ESP32-C3 with external memory (default size TBD, 4MB?)
- Antenna: PCB + IPEX optional (_can ceramic antenna be considered?_)
- USB connector: Type-C (no external USB-to-UART)
- Pin layout: Similar/same as Adafruit Feather (breadboard and shield friendly)

**Peripherals**

- Temperature and Humidity: SHTC3 (https://crates.io/crates/shtcx)
- IMU: BHI160 but no crate for this sensor (BMI160 is NRND).
- PMIC/Charger: MCP73831 (_JST-PH 2.0 SMD connector?_)
- SDCard slot (define MPN)
- LED (regular, not addressable) _Define color!_
- BOOT and RESET button

We need to keep in mind that this board will be used for Rust demonstrations but not limited to. For this reason, we need to keep this board with minimal peripherals that can be used for everyone, especially during workshops and training.
The possibility to add shields could be used for specific demos, like LCD, BLDC, audio, etc.

The main reason for using the bare SoC is due to hackability. We can change the flash size or change the SoC to the ESP32-C3Fx4 (no shield removal needed).

Any suggestions/feedback are welcome!

**Comment by pedrominatel** [comment]
Hi everyone!

In order to accelerate our board development, we'll keep this brainstorming open until Sunday, Nov 28th for new suggestions, recommendations, and specifications regarding the minimum requirements for this board.

We are also discussing about the pin layout here: https://github.com/esp-rs/esp-rust-board/discussions/7

Suggestions for BMI160 alternatives are also welcome!

- BHI160?
- FIS1100?


Thanks!

**Reply by MabezDev** [reply]
I have no real preference, writing a driver should be pretty straight forward for either. I'd say in this climate, whatever we can get our hands on works :smile:.
